servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt

```diff
@@ -16,37 +16,153 @@
+        graphQL.instrumentation?.let { instrumentation ->
+            when {
+                instrumentation is ChainedInstrumentation -> {
+                    instrumentation.instrumentations
+                        .firstOrNull(Instrumentation::isBatchDataLoaderInstrumentation)
+                        ?.javaClass
+                }
+                instrumentation.isBatchDataLoaderInstrumentation() -> instrumentation.javaClass
+                else -> null
+            }
+        }
+
     /**
      * Execute a GraphQL request in a non-blocking fashion.
      * This should only be used for queries and mutations.
      * Subscriptions require more specific server logic and will need to be handled separately.
      */
-    open suspend fun executeRequest(request: GraphQLRequest, context: GraphQLContext? = null, graphQLContext: Map<*, Any> = emptyMap<Any, Any>()): GraphQLResponse<*> {
-        // We should generate a new registry for every request
-        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
-        val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+    open suspend fun executeRequest(
+        graphQLRequest: GraphQLServerRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>()
+    ): GraphQLServerResponse =
+        when (graphQLRequest) {
+            is GraphQLRequest -> execute(graphQLRequest, context, graphQLContext)
+            is GraphQLBatchRequest -> {
+                if (graphQLRequest.containsMutation()) {
+                    executeBatchSequentially(graphQLRequest, context, graphQLContext)
+                } else {
+                    executeBatchConcurrently(graphQLRequest, context, graphQLContext)
+                }
+            }
+        }
 
-        return try {
-            graphQL.executeAsync(executionInput).await().toGraphQLResponse()
+    private suspend fun execute(
+        graphQLRequest: GraphQLRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>(),
+        dataLoaderRegistry: KotlinDataLoaderRegistry? = dataLoaderRegistryFactory?.generate()
+    ): GraphQLResponse<*> =
+        try {
+            graphQL.executeAsync(
+                graphQLRequest.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+            ).await().toGraphQLResponse()
         } catch (exception: Exception) {
             val error = exception.toGraphQLError()
             GraphQLResponse<Any?>(errors = listOf(error.toGraphQLKotlinType()))
         }
+
+    private suspend fun executeBatchSequentially(
+        batchRequest: GraphQLBatchRequest,
+        context: GraphQLContext?,
+        graphQLContext: Map<*, Any>
+    ): GraphQLBatchResponse {
+        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
+        // only 1 operation will be executed at a time, still we can use batching and deduplication
```

Comment:
should this be applied to all then? (i.e. even if it is a single call)

Reply:
when executing batch sequentially only 1 operation will be executed at a time
the instrumentations have this mechanism in place to check when to dispatch, and for that it needs to know how
many operations will be executed concurrently, when doing it sequentially will always be 1

Reply:
Yes it would be applied for single op. I thought that new instrumentation still provided some additional functionality over default dd instrumentation.

Reply:
if clients configure server to use `DataLoaderSyncExecutionExhaustedInstrumentation` they will get the [additional functionality of sync exhaustion](cutions/graphql-kotlin-dataloader-instrumentation/src/test/kotlin/com/expediagroup/graphql/dataloader/instrumentation/syncexhaustion/DataLoaderSyncExecutionExhaustedInstrumentationTest.kt#L229)

Reply:
sure but that also implies the usage of the dataloader through the graphql context?

Reply:
yes, unfortunately, lets see if we can add this instrumentations to graphql-java and always rely on the registry to be placed in the context instead of the execution input